File: frontend/src/services/laporanService.ts
// ============================================
// LAPORAN SERVICE
// KASE AI
// ============================================

export interface LaporanSummary {
  total_penjualan: number;
  total_pembelian: number;
  laba: number;
  jumlah_transaksi: number;

  [key: string]: unknown;
}

export interface LaporanPenjualan {
  id?: number;
  no_transaksi?: string;
  tanggal?: string;
  pelanggan?: string;
  total?: number;
  status?: string;

  [key: string]: unknown;
}

export interface LaporanPembelian {
  id?: number;
  no_transaksi?: string;
  no_faktur?: string;
  tanggal?: string;
  supplier?: string;
  total?: number;
  status?: string;

  [key: string]: unknown;
}


// ============================================
// KONFIGURASI API
// ============================================

const API_BASE_URL =
  import.meta.env.VITE_API_URL;

const API_PREFIX =
  `${API_BASE_URL}/api/v1`;

const LAPORAN_URL =
  `${API_PREFIX}/laporan`;


// ============================================
// REQUEST HELPER
// ============================================

async function request<T>(
  url: string
): Promise<T> {

  const token =
    localStorage.getItem(
      "access_token"
    );

  const response = await fetch(
    url,
    {
      method: "GET",

      headers: {
        Accept:
          "application/json",

        ...(token
          ? {
              Authorization:
                `Bearer ${token}`,
            }
          : {}),
      },
    }
  );


  const contentType =
    response.headers.get(
      "content-type"
    ) || "";


  if (!response.ok) {

    const text =
      await response.text();

    throw new Error(
      `API Laporan gagal ` +
      `(${response.status}): ` +
      `${text.slice(0, 300)}`
    );
  }


  if (
    !contentType.includes(
      "application/json"
    )
  ) {

    const text =
      await response.text();

    throw new Error(
      "API Laporan tidak " +
      "mengembalikan JSON. " +
      `Response: ${text.slice(0, 300)}`
    );
  }


  return response.json();
}


// ============================================
// SUMMARY
// GET /api/v1/laporan/summary
// ============================================

export async function getSummary() {

  return request<LaporanSummary>(
    `${LAPORAN_URL}/summary`
  );
}


// ============================================
// PENJUALAN
// GET /api/v1/laporan/penjualan
// ============================================

export async function getPenjualan() {

  return request<LaporanPenjualan[]>(
    `${LAPORAN_URL}/penjualan`
  );
}


// ============================================
// PEMBELIAN
// GET /api/v1/laporan/pembelian
// ============================================

export async function getPembelian() {

  return request<LaporanPembelian[]>(
    `${LAPORAN_URL}/pembelian`
  );
}


// ============================================
// FILTER PENJUALAN
//
// GET /api/v1/laporan/penjualan/filter
// ============================================

export interface FilterPenjualanParams {
  tanggal_awal?: string;
  tanggal_akhir?: string;
  kategori_id?: number;
  supplier_id?: number;
  pelanggan_id?: number;
  barang_id?: number;
  kasir_id?: number;
  metode_bayar?: string;
  status?: string;
  keyword?: string;
  sort_by?: string;
  sort_order?: string;
  page?: number;
  limit?: number;
}


export async function filterPenjualan(
  params: FilterPenjualanParams = {}
) {

  const query =
    new URLSearchParams();


  if (params.tanggal_awal) {
    query.set(
      "tanggal_awal",
      params.tanggal_awal
    );
  }


  if (params.tanggal_akhir) {
    query.set(
      "tanggal_akhir",
      params.tanggal_akhir
    );
  }


  if (
    params.kategori_id !== undefined
  ) {
    query.set(
      "kategori_id",
      String(params.kategori_id)
    );
  }


  if (
    params.supplier_id !== undefined
  ) {
    query.set(
      "supplier_id",
      String(params.supplier_id)
    );
  }


  if (
    params.pelanggan_id !== undefined
  ) {
    query.set(
      "pelanggan_id",
      String(params.pelanggan_id)
    );
  }


  if (
    params.barang_id !== undefined
  ) {
    query.set(
      "barang_id",
      String(params.barang_id)
    );
  }


  if (
    params.kasir_id !== undefined
  ) {
    query.set(
      "kasir_id",
      String(params.kasir_id)
    );
  }


  if (params.metode_bayar) {
    query.set(
      "metode_bayar",
      params.metode_bayar
    );
  }


  if (params.status) {
    query.set(
      "status",
      params.status
    );
  }


  if (params.keyword) {
    query.set(
      "keyword",
      params.keyword
    );
  }


  query.set(
    "sort_by",
    params.sort_by ||
      "created_at"
  );


  query.set(
    "sort_order",
    params.sort_order ||
      "DESC"
  );


  query.set(
    "page",
    String(params.page || 1)
  );


  query.set(
    "limit",
    String(params.limit || 20)
  );


  const queryString =
    query.toString();


  const url =
    `${LAPORAN_URL}` +
    `/penjualan/filter` +
    `?${queryString}`;


  return request(
    url
  );
}


// ============================================
// RETUR PENJUALAN
// ============================================

export async function getReturPenjualan() {

  return request(
    `${LAPORAN_URL}/retur-penjualan`
  );
}


// ============================================
// RETUR PEMBELIAN
// ============================================

export async function getReturPembelian() {

  return request(
    `${LAPORAN_URL}/retur-pembelian`
  );
}


// ============================================
// STOK
// ============================================

export async function getStok() {

  return request(
    `${LAPORAN_URL}/stok`
  );
}


// ============================================
// MUTASI STOK
// ============================================

export async function getMutasiStok() {

  return request(
    `${LAPORAN_URL}/mutasi-stok`
  );
}


// ============================================
// PRODUK TERLARIS
// ============================================

export async function getProdukTerlaris(
  limit = 10
) {

  return request(
    `${LAPORAN_URL}` +
    `/produk-terlaris?limit=${limit}`
  );
}


// ============================================
// PELANGGAN
// ============================================

export async function getPelanggan() {

  return request(
    `${LAPORAN_URL}/pelanggan`
  );
}


// ============================================
// SUPPLIER
// ============================================

export async function getSupplier() {

  return request(
    `${LAPORAN_URL}/supplier`
  );
}


// ============================================
// LABA RUGI
// ============================================

export async function getLabaRugi() {

  return request(
    `${LAPORAN_URL}/laba-rugi`
  );
}


// ============================================
// NILAI PERSEDIAAN
// ============================================

export async function getNilaiPersediaan() {

  return request(
    `${LAPORAN_URL}/nilai-persediaan`
  );
}


// ============================================
// EXPORT PDF PENJUALAN
// ============================================

export function getPenjualanPdfUrl() {

  return (
    `${LAPORAN_URL}` +
    `/penjualan/pdf`
  );
}


// ============================================
// EXPORT EXCEL PENJUALAN
// ============================================

export function getPenjualanExcelUrl() {

  return (
    `${LAPORAN_URL}` +
    `/penjualan/excel`
  );
}


// ============================================
// SERVICE OBJECT
// ============================================

export const laporanService = {

  getSummary,

  getPenjualan,

  getPembelian,

  filterPenjualan,

  getReturPenjualan,

  getReturPembelian,

  getStok,

  getMutasiStok,

  getProdukTerlaris,

  getPelanggan,

  getSupplier,

  getLabaRugi,

  getNilaiPersediaan,

  getPenjualanPdfUrl,

  getPenjualanExcelUrl,
};


export default laporanService;

File: backend/app/services/laporan_service.py
"""
Laporan Service
===============

Business logic untuk modul laporan.
"""

from app.repositories.laporan_repository import LaporanRepository


class LaporanService:
    """
    Service laporan.
    """

    def __init__(self, db):
        self.repository = LaporanRepository(db)

    # ==========================================
    # Dashboard
    # ==========================================

    def get_dashboard_summary(self):
        return self.repository.get_dashboard_summary()

    # ==========================================
    # Penjualan
    # ==========================================

    def get_penjualan(self):
        return self.repository.get_penjualan()

    # ==========================================
    # Pembelian
    # ==========================================

    def get_pembelian(self):
        return self.repository.get_pembelian()

    # ==========================================
    # Retur Penjualan
    # ==========================================

    def get_retur_penjualan(self):
        return self.repository.get_retur_penjualan()

    # ==========================================
    # Retur Pembelian
    # ==========================================

    def get_retur_pembelian(self):
        return self.repository.get_retur_pembelian()

    # ==========================================
    # Persediaan
    # ==========================================

    def get_stock(self):
        return self.repository.get_stock()

    # ==========================================
    # Mutasi Stok
    # ==========================================

    def get_stock_mutation(self):
        return self.repository.get_stock_mutation()

    # ==========================================
    # Produk Terlaris
    # ==========================================

    def get_top_product(self, limit: int = 10):
        return self.repository.get_top_product(limit)

    # ==========================================
    # Pelanggan
    # ==========================================

    def get_customer_report(self):
        return self.repository.get_customer_report()

    # ==========================================
    # Supplier
    # ==========================================

    def get_supplier_report(self):
        return self.repository.get_supplier_report()

    # ==========================================
    # Laba Rugi
    # ==========================================

    def get_profit_loss(
        self,
        tanggal_awal=None,
        tanggal_akhir=None
    ):
        return self.repository.get_profit_loss(
            tanggal_awal=tanggal_awal,
            tanggal_akhir=tanggal_akhir
        )

    # ==========================================
    # Nilai Persediaan
    # ==========================================

    def get_inventory_value(self):
        return self.repository.get_inventory_value()

    # ==========================================
    # Ringkasan Lengkap
    # ==========================================

    def get_summary(self):
        return {
            "dashboard": self.get_dashboard_summary(),
            "profit_loss": self.get_profit_loss(),
            "inventory": self.get_inventory_value(),
            "top_product": self.get_top_product(),
        }

    # ==========================================
    # Filter Penjualan
    # ==========================================

    def filter_penjualan(
        self,
        tanggal_awal=None,
        tanggal_akhir=None,
        kategori_id=None,
        supplier_id=None,
        pelanggan_id=None,
        barang_id=None,
        kasir_id=None,
        metode_bayar=None,
        status=None,
        keyword=None,
        sort_by="created_at",
        sort_order="DESC",
        page=1,
        limit=20
    ):
        """
        Filter laporan penjualan.
        """

        # --------------------------------------
        # Validasi Pagination
        # --------------------------------------

        if page is None or page < 1:
            page = 1

        if limit is None or limit < 1:
            limit = 20

        if limit > 100:
            limit = 100

        # --------------------------------------
        # Validasi Sorting
        # --------------------------------------

        allowed_sort = {
            "created_at",
            "tanggal",
            "total",
            "no_faktur",
        }

        if sort_by not in allowed_sort:
            sort_by = "created_at"

        if not sort_order:
            sort_order = "DESC"

        sort_order = sort_order.upper()

        if sort_order not in ("ASC", "DESC"):
            sort_order = "DESC"

        # --------------------------------------
        # Kirim ke Repository
        # --------------------------------------

        return self.repository.filter_penjualan(
            tanggal_awal=tanggal_awal,
            tanggal_akhir=tanggal_akhir,
            kategori_id=kategori_id,
            supplier_id=supplier_id,
            pelanggan_id=pelanggan_id,
            barang_id=barang_id,
            kasir_id=kasir_id,
            metode_bayar=metode_bayar,
            status=status,
            keyword=keyword,
            sort_by=sort_by,
            sort_order=sort_order,
            page=page,
            limit=limit,
        )